go-controller/pkg/ovn/secondary_localnet_network_controller.go

```diff
@@ -360,11 +403,82 @@ func (oc *SecondaryLocalnetNetworkController) newRetryFramework(
 }
 
 func (oc *SecondaryLocalnetNetworkController) localnetPortNetworkNameOptions() map[string]string {
-	localnetLSPOptions := map[string]string{
-		"network_name": oc.GetNetworkName(),
+	return map[string]string{
+		"network_name": oc.findNetworkName(),
+	}
+}
+
+func (oc *SecondaryLocalnetNetworkController) checkBridgeMapping(obj interface{}) error {
+	if !config.OVNKubernetesFeature.EnableInterconnect {
+		return nil
+	}
+
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		return fmt.Errorf("failed casting %T object to *knet.Pod", obj)
+	}
+
+	if !util.PodScheduled(pod) {
+		return nil
+	}
+
+	if oc.controllerNodeName != pod.Spec.NodeName {
+		return nil
+	}
+
+	isNADKnown, err := oc.isPodAttachedToKnownNAD(pod)
+	if err != nil {
+		return err
+	}
+	if !isNADKnown {
+		return nil
 	}
+
+	return oc.networkMappedToBridge(pod)
+}
+
+func (oc *SecondaryLocalnetNetworkController) isPodAttachedToKnownNAD(pod *corev1.Pod) (bool, error) {
+	networks, err := util.GetK8sPodAllNetworkSelections(pod)
+	if err != nil {
+		return false, err
+	}
+
+	nInfo := oc.GetNetInfo()
+	for _, network := range networks {
+		nadName := util.GetNADName(network.Namespace, network.Name)
+		if nInfo.HasNAD(nadName) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (oc *SecondaryLocalnetNetworkController) networkMappedToBridge(pod *corev1.Pod) error {
+	openvSwitch, err := ovsops.GetOpenvSwitch(oc.ovsClient)
+	if err != nil {
+		return err
+	}
+
+	ovnBridgeMappings := openvSwitch.ExternalIDs["ovn-bridge-mappings"]
+
+	networkName := oc.findNetworkName()
+	bridgeMappings := strings.Split(ovnBridgeMappings, ",")
+	for _, bridgeMapping := range bridgeMappings {
+		networkBridgeAssociation := strings.Split(bridgeMapping, ":")
+		if len(networkBridgeAssociation) == 2 && networkBridgeAssociation[0] == networkName {
+			return nil
+		}
+	}
+
+	err = fmt.Errorf("failed to find bridge mapping for network: %q on node %q; Current ovn-bridge-mappings: %q", networkName, oc.controllerNodeName, ovnBridgeMappings)
+	oc.recordPodErrorEvent(pod, err)
```

Comment:
So if I get this right, the only thing we pretend to do with all this is report an event? We are not blocking or affecting the status of the pod otherwise? If the bridge mapping is not there as reported, wouldn't the CNI side of things fail anyway? Kinda interest on what the global scenario looks like.

Reply:
Right @jcaamano .

It could make sense to flat out crashloop the pod - but we plan in the future to impact scheduling. Meaning, the workload would only be scheduled on a node featuring the required mapping.

This would be a first step, to show the users why their pod's network is broken.

Reply:
Interested in `If the bridge mapping is not there as reported, wouldn't the CNI side of things fail anyway?` also

Reply:
no, it doesn't fail. The localnet port will be created, but the patch port will not, which will black-hole inter-node traffic, while working for workloads scheduled in the same node.

Reply:
What prevents us to do exactly the same check on the CNI side which would be more straight forward because everything its synchronous there? What are we missing?

Reply:
Do we even have access to OVS from the CNI side of things ?

Reply:
@oshoval would you remind me if we actually evaluated doing this from the CNI side ? 

I don't have anything recorded on my notes, and I'm afraid to mix this up w/ the ipv6 story.

Reply:
dont fully remember, if i will recall i will update

but on the CNI, unless the following info is missing:
We will need to find the network that is used by the nads pointed by the NetworkSelectionElements, and to check if this one is a local net

Note, i think i did we dont just report an event but also fails the creation, but will need to double check it if desired

Reply:
IIUC you can use the `networkManager` [here](https://github.com/ovn-kubernetes/ovn-kubernetes/blob/9ea5b14aa3b2dcf2e6ebf36bbf3330337cf262e1/go-controller/pkg/cni/cni.go#L118) to get the NetInfo, from which you can access the topology.

Reply:
> Note about https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5000/files#r2000973063, i think i did we dont just report an event but also fails the creation, but will need to double check it if desired

Please check ! This one's important.

Reply:
Regarding the architecture of this suggestion, doing it in the CNI will pretty much involve adding code common to all the type of networks, which will surely add a performance penalty (... for all types of networks). I think we are using listers, hence, it shouldn't be much, but still ... it's something. At the very least - IMHO - we need to make sure we're not impacting the default cluster network.

Furthermore, the current proposal, which only touches localnet specific code. IMHO, this suggestion is cleaner, and potentially performs better. (I am aware that it's hard to argue for / against without figures ...)

Not exactly sure implementing it in the CNI is more straightforward, but I can see a tradeoff. 
@jcaamano let us know if you really want us to explore this path further.

Reply:
> This implementation here is racey so not functionally correct. If we are not aware of the NAD yet, we might not report the event when we should have. If we are not aware of the bridge mapping yet, we might report the event when we shouldn't have. Once we report, we can't unreport.

If there is no NAD, the pod wont get scheduled because other things and there is retry isnt it ? (i think i saw a retry, we can even try to create the mapping after the event and see if the pod self heal, maybe tried already)
If there is no bridge mapping yet, it is marginal that can also happen on CNI, and there is retry

Reply:
I remember we found that doing it at CNI was quite involved since we have to add listeners to the CNI server, although it's true that it will be clear to the user (no ovn mapping, no pod).

Since this PR is quite old maybe now things has improve and we need less plubmbing and CNI server ?

Reply:
> Please check ! This one's important.

https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5000/files#diff-87fd62ad4d80ef1e2da6f02a7bdc5e4910bd30b1565e1042feba949dfdd23772R117
from what i remember and as we see here 
error + event was the requirements (at least back then of course)

Reply:
> If there is no NAD, the pod wont get scheduled

But the important thing here is the order in which different components become aware of things, not he order in which they actually happen. So the pod can be scheduled, but ourselves may not yet be aware of the existence of the NAD.

> we have to add listeners to the CNI server

Can you elaborate? I don't see the need, perhaps I am missing something.

> https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5000/files#diff-87fd62ad4d80ef1e2da6f02a7bdc5e4910bd30b1565e1042feba949dfdd23772R117

So the only effect that has is recording the event up to 15 times or so which are our maximum retries. This is not failing anything else since we are not doing anything else. Meaning, not failing anything from the user perspective.

Reply:
sure ! right. I'm on a roll today :sweat_smile: 

To sum up: we get access to the topology attribute for free, from the configuration. There's no performance impacts of doing this in the CNI side. As a bonus, this means that we don't have to reconcile NADs at all.

@oshoval I think we should at least try to do this on the CNI side of things.